src/Core/Services/Implementations/OrganizationService.cs

```diff
@@ -184,187 +184,248 @@ public async Task ReinstateSubscriptionAsync(Guid organizationId)
+                                                  "Disable your Password Reset policy.");
+                }
             }
-        }
 
-        if (!newPlan.HasCustomPermissions && organization.UseCustomPermissions)
-        {
-            var organizationCustomUsers =
-                await _organizationUserRepository.GetManyByOrganizationAsync(organization.Id,
-                    OrganizationUserType.Custom);
-            if (organizationCustomUsers.Any())
+            if (!newPlan.HasScim && organization.UseScim)
             {
-                throw new BadRequestException("Your new plan does not allow the Custom Permissions feature. " +
-                                              "Disable your Custom Permissions configuration.");
+                var scimConnections = await _organizationConnectionRepository.GetByOrganizationIdTypeAsync(
+                    organization.Id,
+                    OrganizationConnectionType.Scim);
+                if (scimConnections != null && scimConnections.Any(c => c.GetConfig<ScimConfig>()?.Enabled == true))
+                {
+                    throw new BadRequestException("Your new plan does not allow the SCIM feature. " +
+                                                  "Disable your SCIM configuration.");
+                }
             }
-        }
 
-        // TODO: Check storage?
+            if (!newPlan.HasCustomPermissions && organization.UseCustomPermissions)
+            {
+                var organizationCustomUsers =
+                    await _organizationUserRepository.GetManyByOrganizationAsync(organization.Id,
+                        OrganizationUserType.Custom);
+                if (organizationCustomUsers.Any())
+                {
+                    throw new BadRequestException("Your new plan does not allow the Custom Permissions feature. " +
+                                                  "Disable your Custom Permissions configuration.");
+                }
+            }
 
-        string paymentIntentClientSecret = null;
-        var success = true;
-        if (string.IsNullOrWhiteSpace(organization.GatewaySubscriptionId))
-        {
-            paymentIntentClientSecret = await _paymentService.UpgradeFreeOrganizationAsync(organization, newPlan,
-                upgrade.AdditionalStorageGb, upgrade.AdditionalSeats, upgrade.PremiumAccessAddon, upgrade.TaxInfo);
-            success = string.IsNullOrWhiteSpace(paymentIntentClientSecret);
-        }
-        else
-        {
-            // TODO: Update existing sub
-            throw new BadRequestException("You can only upgrade from the free plan. Contact support.");
+            // TODO: Check storage?
+
+
+            if (string.IsNullOrWhiteSpace(organization.GatewaySubscriptionId))
+            {
+                var organizationUpgradePlan = useSecretsManager
+                    ? newPlans
+                    : newPlans.Where(x => x.BitwardenProduct == BitwardenProductType.PasswordManager).Take(1).ToList();
```

Comment:
I may be repeating myself, but just in case: you should do this filtering when you declare `newPlans`, so that it's accurate from the start (that is, it only contains the plans the organization has actually signed up for). Instead of including both plans and then only filtering them later.

Reply:
Worked on